### DISCUSSION TITLE
Configure FreemarkerViewRenderer

### DISCUSSION THREAD

**Discussion by JKomoroski** [dropwizard/dropwizard#8430] [Q&A] [answered]
I haven't been able to find a way to configure the Freemarker feature version `freemarker.template.Version` and the caching behavior for the FreemarkerViewRenderer.

Since the FreemarkerViewRenderer registers itself via the service loader mechanism, the default no-args constructor appears to be called . This results in the view renderer always having the default version of `2.3.0`.

Is there a way to configure that version without extending the class?

Example:
```
import com.google.auto.service.AutoService;
import freemarker.template.Version;
import io.dropwizard.views.common.ViewRenderer;
import io.dropwizard.views.freemarker.FreemarkerViewRenderer;

@AutoService(ViewRenderer.class)
public class CustomFVR extends FreemarkerViewRenderer implements ViewRenderer {

    public CustomFVR() {
        super(new Version(2, 3, 32));
    }

    @Override
    public String getConfigurationKey() {
        return "customfreemarker";
    }
}
```

Also, I don't think the caching behavior for FreemarkerViews can be configured as currently designed. The mustashe implementation seems to have special logic for that behavior, but I might have missed a setting for that one.

**Comment by joschi** [comment, marked as answer]
You can specify the version with the `incompatible_improvements` configuration setting for Freemarker, see [`Configuration#setSetting(String, String)`](https://freemarker.apache.org/docs/api/freemarker/template/Configuration.html#setSetting-java.lang.String-java.lang.String-):
```yaml
# configuration.yaml
views:
  freemarker:
    incompatible_improvements: "2.3.32"
```

> "incompatible_improvements": See [setIncompatibleImprovements(Version)](https://freemarker.apache.org/docs/api/freemarker/template/Configuration.html#setIncompatibleImprovements-freemarker.template.Version-).
> String value: version number like 2.3.20.

> Also, I don't think the caching behavior for FreemarkerViews can be configured as currently designed.

Which behavior do you mean exactly?

> <li><p><code>"cache_storage"</code>:
>        See <a href="https://freemarker.apache.org/docs/api/freemarker/template/Configuration.html#setCacheStorage-freemarker.cache.CacheStorage-"><code>setCacheStorage(freemarker.cache.CacheStorage)</code></a>.

**Reply by JKomoroski** [reply]
> You can specify the version with the incompatible_improvements configuration setting for Freemarker, see [Configuration#setSetting(String, String)](https://freemarker.apache.org/docs/api/freemarker/template/Configuration.html#setSetting-java.lang.String-java.lang.String-):

:rofl: What a nice way of telling me to RTFM. I totally misunderstood what was in the documentation. Thank you.

As for the caching, I did some debugging now understand the way that the freemarker Configurations are cached per view, and Freemarker Configuration storage cache be disabled with `cache_storage: freemarker.cache.NullCacheStorage()`. (writing this here so I will find it next time)

Thanks again joschi.
